src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -0,0 +1,866 @@
+            }
+
+            if (isCanceled)
+            {
+                _readingState.BeginTentative();
+            }
+            else
+            {
+                _readingState.Begin();
+            }
+        }
+
+        internal bool IsFlushAsyncCompleted => _writerAwaitable.IsCompleted;
+
+        internal FlushResult GetFlushAsyncResult()
+        {
+            var result = new FlushResult();
+            lock (_sync)
+            {
+                if (!_writerAwaitable.IsCompleted)
+                {
+                    ThrowHelper.ThrowInvalidOperationException_GetResultNotCompleted();
+                }
+
+                // Change the state from to be canceled -> observed
+                if (_writerAwaitable.ObserveCancelation())
+                {
+                    result.ResultFlags |= ResultFlags.Canceled;
+                }
+                if (_readerCompletion.IsCompletedOrThrow())
+                {
+                    result.ResultFlags |= ResultFlags.Completed;
+                }
+            }
+
+            return result;
+        }
+
+        internal void OnFlushAsyncCompleted(Action continuation)
+        {
+            Action awaitable;
+            bool doubleCompletion;
+            lock (_sync)
+            {
+                awaitable = _writerAwaitable.OnCompleted(continuation, out doubleCompletion);
+            }
+            if (doubleCompletion)
+            {
+                Reader.Complete(ThrowHelper.CreateInvalidOperationException_NoConcurrentOperation());
+            }
+            TrySchedule(_writerScheduler, awaitable);
+        }
+
+        private void ReaderCancellationRequested()
+        {
+            Action action;
+            lock (_sync)
+            {
+                action = _readerAwaitable.Cancel();
+            }
+            TrySchedule(_readerScheduler, action);
+        }
+
+        private void WriterCancellationRequested()
+        {
+            Action action;
+            lock (_sync)
+            {
+                action = _writerAwaitable.Cancel();
+            }
+            TrySchedule(_writerScheduler, action);
+        }
+
+        /// <summary>
+        /// Gets the <see cref="PipeReader"/> for this pipe.
+        /// </summary>
+        public PipeReader Reader => _reader;
+
+        /// <summary>
+        /// Gets the <see cref="PipeWriter"/> for this pipe.
+        /// </summary>
+        public PipeWriter Writer => _writer;
+
+        /// <summary>
+        /// Resets the pipe
+        /// </summary>
+        public void Reset()
+        {
+            lock (_sync)
+            {
+                if (!_disposed)
+                {
+                    ThrowHelper.ThrowInvalidOperationException_ResetIncompleteReaderWriter();
+                }
+
+                _disposed = false;
+                ResetState();
+            }
+        }
+
+        private sealed class DefaultPipeReader : PipeReader, IPipeAwaiter<ReadResult>
```

Comment:
These are non-trivial-sized nested classes in a file that's already quite large. Please move these two to separate files.